src/bosh-dns/dns/server/handlers/failover_recursor_pool.go

```diff
@@ -89,26 +112,45 @@ func newSmartFailoverRecursorPool(recursors []string, logger logger.Logger) Recu
 		preferredRecursorIndex: 0,
 		logger:                 logger,
 		logTag:                 logTag,
+		recursorRetrySettings:  recursorSettings,
 	}
 }
 
 func (q *serialFailoverRecursorPool) PerformStrategically(work func(string) error) error {
 	for _, r := range q.recursors {
-		if err := work(r); err == nil {
+		if err := performWithRetryLogic(work, r, q.recursorRetrySettings.retryCount, q.logTag, q.logger); err == nil {
 			return nil
 		}
 	}
-
 	return ErrNoRecursorResponse
 }
 
+func performWithRetryLogic(work func(string) error, recursor string, retryCount int, logTag string, log logger.Logger) (err error) {
+	for ret := -1; ret < retryCount; ret++ {
+		err = work(recursor)
+		if err == nil {
+			return err
+		}
+		if _, ok := err.(net.Error); !ok {
+			log.Debug(logTag, fmt.Sprintf("dns request error %v no retry - %v\n", err, recursor))
+			return err
+		}
+		log.Error(logTag, fmt.Sprintf("dns request network error %s retry [%d/%d] for recoursor %s \n", err.(net.Error), ret+1, retryCount, recursor))
```

Comment:
Can we switch this to debug? Given that this feature is to optimize packet drop behavior, this doesn't seem like an error worth especially calling out.

Reply:
Yes, fine for me

Reply:
> We're deviating from my understanding of [this doc](https://tools.ietf.org/html/rfc1536) now with a retry on the same recursor. I think that's probably fine as one advantage I can see is that it will help call out down recursors vs just recursors that don't have the answer.

Yes, there is no part at the **rfc** that matches 100% so we have to go some small things different here.

Think the implementation was fine at this point, because we are using UDP and have a **fast retransmission** for the lost packages (thats the most important thing).

The retry is also not the "default" way... so in most of the cases the retry will never be called so we don't loose any time here and didn't add to much complexity.

Reply:
It would be undiscoverable how many retries happened unless debug logging is turned on, which we would not do in production.

Reply:
Would info be more appropriate then? I tend to think of errors as indicators of a failing system and since we're building in some resiliency to that, I think the error below is the real indicator. 

Also in the 0 retry default case, we'd get two errors for the same packet loss which could be confusing?